pkg/controller/projects/accesstokens/controller.go

```diff
@@ -0,0 +1,237 @@
+func SetupAccessToken(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.AccessTokenKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.AccessToken{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.AccessTokenGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), newGitlabClientFn: newAccessTokenClient}),
+			managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient())),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connector struct {
+	kube              client.Client
+	newGitlabClientFn func(cfg clients.Config) gitlab.ProjectAccessTokensService
+}
+
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1alpha1.AccessToken)
+	if !ok {
+		return nil, errors.New(errNotAccessToken)
+	}
+	cfg, err := clients.GetConfig(ctx, c.kube, cr)
+	if err != nil {
+		return nil, err
+	}
+	return &external{kube: c.kube, client: c.newGitlabClientFn(*cfg)}, nil
+}
+
+type external struct {
+	kube   client.Client
+	client gitlab.ProjectAccessTokensService
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.AccessToken)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotAccessToken)
+	}
+
+	externalName := meta.GetExternalName(cr)
+	if externalName == "" {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+
+	projectAccessTokenID, err := strconv.Atoi(externalName)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.New(errNotAccessToken)
+	}
+
+	listProjectAccessTokensOptions := gitlab.ListProjectAccessTokensOptions{}
+	accesstokenArr, _, err := e.client.ListProjectAccessTokens(*cr.Spec.ForProvider.ProjectID, &listProjectAccessTokensOptions)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(isErrorProjectAccessTokenNotFound, err), errGetFailed)
+	}
+
+	dt := findAccessToken(projectAccessTokenID, accesstokenArr)
+
+	if dt == nil {
+		return managed.ExternalObservation{}, nil
+	}
+
+	current := cr.Spec.ForProvider.DeepCopy()
+	lateInitializeProjectAccessToken(&cr.Spec.ForProvider, dt)
+
+	cr.Status.AtProvider = v1alpha1.AccessTokenObservation{}
+	cr.Status.SetConditions(xpv1.Available())
+
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        true,
+		ResourceLateInitialized: !cmp.Equal(current, &cr.Spec.ForProvider),
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.AccessToken)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotAccessToken)
+	}
+
+	dt, _, err := e.client.CreateProjectAccessToken(
+		*cr.Spec.ForProvider.ProjectID,
```

Comment:
possible nil deref here: `cr.Spec.ForProvider.ProjectID` is an optional field. You'd have to check if it defined first and error out if that's the case.